src/ossl/ecdh.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright 2023 - 2024 Simo Sorce, Jakub Jelen
+// See LICENSE.txt file for terms
+
+use core::ffi::{c_char, c_int, c_uint};
+use std::borrow::Cow;
+
+use crate::attribute::CkAttrs;
+use crate::bytes_to_vec;
+use crate::ecc::*;
+use crate::error::Result;
+use crate::interface::*;
+use crate::mechanism::*;
+use crate::object::{default_key_attributes, Object, ObjectFactories};
+use crate::ossl::bindings::*;
+use crate::ossl::common::*;
+#[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+use crate::ossl::ec_montgomery::{
+    is_montgomery_curve, make_ec_montgomery_public_key,
+    make_output_length_from_montgomery_obj,
+    montgomery_object_to_ecc_private_key,
+};
+#[cfg(feature = "ecc")]
+use crate::ossl::ecc::{
+    ecdsa_object_to_ecc_private_key, get_curve_name_from_obj,
+    make_ecdsa_public_key, make_output_length_from_ecdsa_obj,
+};
+
+fn kdf_type_to_hash_mech(mech: CK_EC_KDF_TYPE) -> Result<CK_MECHANISM_TYPE> {
+    match mech {
+        CKD_SHA1_KDF => Ok(CKM_SHA_1),
+        CKD_SHA224_KDF => Ok(CKM_SHA224),
+        CKD_SHA256_KDF => Ok(CKM_SHA256),
+        CKD_SHA384_KDF => Ok(CKM_SHA384),
+        CKD_SHA512_KDF => Ok(CKM_SHA512),
+        CKD_SHA3_224_KDF => Ok(CKM_SHA3_224),
+        CKD_SHA3_256_KDF => Ok(CKM_SHA3_256),
+        CKD_SHA3_384_KDF => Ok(CKM_SHA3_384),
+        CKD_SHA3_512_KDF => Ok(CKM_SHA3_512),
+        _ => return Err(CKR_MECHANISM_PARAM_INVALID)?,
+    }
+}
+
+fn make_derive_output_length_from_obj(key: &Object) -> Result<usize> {
+    // Montgomery
+    #[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+    if is_montgomery_curve(key) {
+        return make_output_length_from_montgomery_obj(key);
+    }
+    // ECDSA
+    make_output_length_from_ecdsa_obj(key)
+}
+
+fn derive_object_to_ecc_private_key(key: &Object) -> Result<EvpPkey> {
+    // Montgomery
+    #[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+    if is_montgomery_curve(key) {
+        return montgomery_object_to_ecc_private_key(key);
+    }
+    // ECDSA
+    ecdsa_object_to_ecc_private_key(key)
```

Comment:
I have a feeling that all these functions to create EvpPkey from ecc/ec_montgomery/ etc keys should actually be moved into an ossl/ecc_misc.rs file this way the config directives stay there and return appropriate errors if the key type is not correct.

Currently this code will return bad data if the caller passes in a key object with CKA_KEY_TYPE = CKK_EC_MONTGOMERY but the code compiled out support "ec_montgomery" as ecdsa_object_to_ecc_private_key() does not check that the object provide is of the correct type as far as I can tell and the underlying openssl code could be supporting montgomery keys and return an EvpPkey.

Reply:
Even if these are not moved I think the code needs to be changed to something like:
```
let key_type = key.get_attr_as_ulong(CKA_KEY_TYPE)?;
match key_type {
    CKK_ECC => ecdsa_object_to_ecc_private_key(key),
    CKK_EC_MONTGOMERY => montgomery_object_to_ecc_private_key(key),
    _ => Err(CKR_KEY_TYPE_INVALID)?,
}
```

Reply:
How elementary! I knew there will be some better way, but forgot about the key type!

About moving the functions to ecc_misc, I am worried that it would require moving too many curve specific stuff in there (the ossl name used, for ecdsa the group name extraction ...).